soco/music_services/music_service.py

```diff
@@ -0,0 +1,780 @@
+        #         ContainerType="MService"
+        #         Capabilities="513"
+        #         MaxMessagingChars="0">
+        #         <Policy Auth="Anonymous" PollInterval="30" />
+        #         <Presentation>
+        #             <Strings
+        #                 Version="1"
+        #                 Uri="https:...string_table.xml" />
+        #             <PresentationMap Version="2"
+        #                 Uri="https://...presentation_map.xml" />
+        #         </Presentation>
+        #     </Service>
+        # ...
+        # </ Services>
+
+        # Ideally, the search path should be './/Service' to find Service
+        # elements at any level, but Python 2.6 breaks with this if Service
+        # is a child of the current element. Since 'Service' works here, we use
+        # that instead
+        services = root.findall('Service')
+        for service in services:
+            result_value = service.attrib.copy()
+            name = service.get('Name')
+            result_value['Name'] = name
+            auth_element = (service.find('Policy'))
+            auth = auth_element.attrib
+            result_value.update(auth)
+            presentation_element = (service.find('.//PresentationMap'))
+            if presentation_element is not None:
+                result_value['PresentationMapUri'] = \
+                    presentation_element.get('Uri')
+            result_value['ServiceID'] = service.get('Id')
+            # ServiceType is used elsewhere in Sonos, eg to form tokens,
+            # and get_subscribed_music_services() below. It is also the
+            # 'Type' used in account_xml (see above). Its value always
+            # seems to be (ID*256) + 7. Some serviceTypes are also
+            # listed in available_services['AvailableServiceTypeList']
+            # but this does not seem to be comprehensive
+            service_type = str(int(service.get('Id')) * 256 + 7)
+            result_value['ServiceType'] = service_type
+            result[service_type] = result_value
+        # Cache this so we don't need to do it again.
+        cls._music_services_data = result
+        return result
+
+    @classmethod
+    def get_all_music_services_names(cls):
+        """Get a list of the names of all available music services.
+
+        These services have not necessarily been subscribed to.
+
+        Returns:
+            (list): A list of strings
+
+        """
+        return [
+            service['Name'] for service in
+            cls._get_music_services_data().values()
+        ]
+
+    @classmethod
+    def get_subscribed_services_names(cls):
+        """Get a list of the names of all subscribed music services.
+
+        The TuneIn service is always subscribed but will not appear in the list
+
+        Returns:
+            (list): A list of strings
+        """
+        # This is very inefficient - loops within loops within loops, and
+        # many network requests
+        # Optimise it?
```

Comment:
But should be called rather infrequently. For now, I wouldn't bother. Let the performance become problem first.